LEGO1/tgl/d3drm/mesh.cpp

```diff
@@ -0,0 +1,166 @@
+	}
+}
+
+// OFFSET: LEGO1 0x100a3f50
+Result MeshImpl::SetTexture(const Texture* p_texture)
+{
+	IDirect3DRMTexture* texture = p_texture ? static_cast<const TextureImpl*>(p_texture)->ImplementationData() : NULL;
+	return ResultVal(m_data->groupMesh->SetGroupTexture(m_data->groupIndex, texture));
+}
+
+// OFFSET: LEGO1 0x100a3f80
+Result MeshImpl::SetTextureMappingMode(ProjectionType p_projType)
+{
+	if (p_projType == Perspective) {
+		return ResultVal(m_data->groupMesh->SetGroupMapping(m_data->groupIndex, D3DRMMAP_PERSPCORRECT));
+	}
+	else {
+		return ResultVal(m_data->groupMesh->SetGroupMapping(m_data->groupIndex, 0));
+	}
+}
+
+// OFFSET: LEGO1 0x100a3fc0
+Result MeshImpl::SetShadingModel(ShadingModel p_model)
+{
+	D3DRMRENDERQUALITY mode;
+	switch (p_model) {
+	case Wireframe:
+		mode = D3DRMRENDER_WIREFRAME;
+		break;
+	case UnlitFlat:
+		mode = D3DRMRENDER_UNLITFLAT;
+		break;
+	case Flat:
+		mode = D3DRMRENDER_FLAT;
+		break;
+	case Gouraud:
+		mode = D3DRMRENDER_GOURAUD;
+		break;
+	case Phong:
+		mode = D3DRMRENDER_PHONG;
+		break;
+	}
+	return ResultVal(m_data->groupMesh->SetGroupQuality(m_data->groupIndex, mode));
+}
+
+// OFFSET: LEGO1 0x100a4030
+Mesh* MeshImpl::DeepClone(Something* p_mesh)
+{
+	// Create group
+	MeshImpl* newSomething = new MeshImpl();
+	MeshData* data = new MeshData();
+	newSomething->m_data = data;
+
+	// Query information from old group
+	DWORD dataSize;
+	unsigned int vcount, fcount, vperface;
+	m_data->groupMesh->GetGroup(m_data->groupIndex, &vcount, &fcount, &vperface, &dataSize, NULL);
+	unsigned int* faceBuffer = (unsigned int*) malloc(dataSize * sizeof(unsigned int));
+	m_data->groupMesh->GetGroup(m_data->groupIndex, &vcount, &fcount, &vperface, &dataSize, faceBuffer);
+	// We expect vertex to be sized 0x24, checked at start of file.
+	D3DRMVERTEX* vertexBuffer = (D3DRMVERTEX*) malloc(vcount * sizeof(D3DRMVERTEX));
+	m_data->groupMesh->GetVertices(m_data->groupIndex, 0, vcount, vertexBuffer);
+	LPDIRECT3DRMTEXTURE textureRef;
+	m_data->groupMesh->GetGroupTexture(m_data->groupIndex, &textureRef);
+	D3DRMMAPPING mapping = m_data->groupMesh->GetGroupMapping(m_data->groupIndex);
+	D3DRMRENDERQUALITY quality = m_data->groupMesh->GetGroupQuality(m_data->groupIndex);
+	D3DCOLOR color = m_data->groupMesh->GetGroupColor(m_data->groupIndex);
+
+	// Push information to new group
+	SomethingImpl* target = static_cast<SomethingImpl*>(p_mesh);
+	D3DRMGROUPINDEX index;
+	target->ImplementationData()->AddGroup(vcount, fcount, vperface, faceBuffer, &index);
+	newSomething->m_data->groupIndex = index;
+	target->ImplementationData()->SetVertices(index, 0, vcount, vertexBuffer);
+	target->ImplementationData()->SetGroupTexture(index, textureRef);
+	target->ImplementationData()->SetGroupMapping(index, mapping);
+	target->ImplementationData()->SetGroupQuality(index, quality);
+	Result result = ResultVal(target->ImplementationData()->SetGroupColor(index, color));
+
+	// Cleanup
+	if (faceBuffer)
+		free(faceBuffer);
```

Comment:
This should be `delete` (below too)

Reply:
This one could be a new/delete, changed.